// react-router-transition-switch/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _warning = _interopRequireDefault(require("warning"));
var _reactRouter = require("react-router");
var _excluded = ["location"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var TransitionSwitch = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(TransitionSwitch, _React$Component);
  var _super = _createSuper(TransitionSwitch);
  function TransitionSwitch() {
    (0, _classCallCheck2["default"])(this, TransitionSwitch);
    return _super.apply(this, arguments);
  }
  (0, _createClass2["default"])(TransitionSwitch, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      (0, _warning["default"])(!(prevProps.overrideLocation && !this.props.overrideLocation), '<TransitionSwitch> elements should not change from uncontrolled to controlled (or vice versa). You initially used no "location" prop and then provided one on a subsequent render.');
      (0, _warning["default"])(!(!prevProps.overrideLocation && this.props.overrideLocation), '<TransitionSwitch> elements should not change from controlled to uncontrolled (or vice versa). You provided a "location" prop initially but omitted it on a subsequent render.');
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        children = _this$props.children,
        render = _this$props.render,
        component = _this$props.component,
        createKey = _this$props.createKey,
        routeMatch = _this$props.match;
      var location = this.props.overrideLocation || this.props.location;
      var match, element;
      React.Children.forEach(children, function (child) {
        if (match == null && /*#__PURE__*/React.isValidElement(child)) {
          var _child$props = child.props,
            pathProp = _child$props.path,
            exact = _child$props.exact,
            strict = _child$props.strict,
            sensitive = _child$props.sensitive,
            from = _child$props.from;
          var path = pathProp || from;
          element = child;
          match = path ? (0, _reactRouter.matchPath)(location.pathname, {
            path: path,
            exact: exact,
            strict: strict,
            sensitive: sensitive
          }) : routeMatch;
        }
      });
      var routeElement = match ? /*#__PURE__*/React.cloneElement(element, {
        location: location,
        computedMatch: match,
        key: createKey(element, match)
      }) : null;
      var props = {
        location: location,
        match: match,
        children: routeElement
      };
      if (component) return /*#__PURE__*/React.createElement(component, props);
      if (render) return render(props);
      return routeElement;
    }
  }]);
  return TransitionSwitch;
}(React.Component);
(0, _defineProperty2["default"])(TransitionSwitch, "propTypes", {
  children: _propTypes["default"].node,
  location: _propTypes["default"].object,
  createKey: _propTypes["default"].func
});
(0, _defineProperty2["default"])(TransitionSwitch, "defaultProps", {
  createKey: function createKey(child, match) {
    return child.key != null ? child.key : match.url;
  }
});
var TransitionSwitch2 = (0, _reactRouter.withRouter)(TransitionSwitch);
var TransitionSwitch3 = function TransitionSwitch3(_ref) {
  var location = _ref.location,
    props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  return /*#__PURE__*/React.createElement(TransitionSwitch2, (0, _extends2["default"])({
    overrideLocation: location
  }, props));
};
var _default = exports["default"] = TransitionSwitch3;
module.exports = exports.default;
